layer/unittest/trace_event_log_tests.cc

```diff
@@ -27,8 +31,167 @@ using ::testing::IsEmpty;
+                 << "\"";
+        break;
+      }
+      case ValueType::kTimestamp: {
+        json_str << ValueToJsonString(
+            args[i]->cast<TimestampAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kDuration: {
+        json_str << ValueToJsonString(
+            args[i]->cast<DurationAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kBool: {
+        json_str << ValueToJsonString(args[i]->cast<BoolAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kInt64: {
+        json_str << ValueToJsonString(args[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::kString: {
+        json_str << std::quoted(args[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kVectorInt64: {
+        json_str << ValueToJsonString(
+            args[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::kTraceEvent:
+        break;
+    }
+  }
+  json_str << " }";
+}
+
+// Appends the start timestamp and the duration of a `TraceEventAttr` to the
+// given stream.
+void CompleteEventToString(TimestampAttr timestamp,
+                           const TraceEventAttr *trace_event,
+                           std::ostringstream &json_stream) {
+  const std::vector<Attribute *> &args = trace_event->GetArgs();
+  assert(args.size() && "CompleteEvent's args must not be empty.");
+
+  auto it = std::find_if(args.begin(), args.end(), [](Attribute *attr) {
+    return attr->isa<DurationAttr>();
+  });
+  assert(it != args.end() &&
+         "Duration not found in args.\nCompleteEvent's args must contain "
+         "DurationAttr.");
+
+  double duration_ms = (*it)->cast<DurationAttr>()->GetValue().ToMilliseconds();
+  double end_timestamp_ms = ToUnixMillis(timestamp.GetValue());
+  double start_timestamp_ms = end_timestamp_ms - duration_ms;
+  json_stream << ", " << std::quoted("ts") << " : " << std::fixed
+              << start_timestamp_ms << ", " << std::quoted("dur") << " : "
+              << duration_ms;
+}
+
+// Appends the timestamp and the scope of a `TraceEventAttr` to the given
+// stream.
+void InstantEventToString(TimestampAttr timestamp,
+                          const TraceEventAttr *trace_event,
+                          std::ostringstream &json_stream) {
+  const std::vector<Attribute *> &args = trace_event->GetArgs();
+  assert(args.size() && "InstantEvent's args must not be empty.");
+
+  auto it = std::find_if(args.begin(), args.end(), [](Attribute *attr) {
+    return attr->isa<StringAttr>() &&
+           attr->GetName() == std::string_view("scope");
+  });
+  assert(it != args.end() && "Scope not found in args.");
+
+  const std::string &scope = (*it)->cast<StringAttr>()->GetValue();
+  assert((scope == std::string_view("g") || scope == std::string_view("p") ||
+          scope == std::string_view("t")) &&
+         "Invalid scope. Scope must be \"g\", \"p\", or \"t\".");
```

Comment:
nit: Hmm, should we make `StringAttr::GetValue` return a `string_view` instead of `char *`? The underlying value can still be `const char*`